### ISSUE TITLE
`mutual` blocks should be functionally equivalent to new-style implicit mutual recursion

### ISSUE THREAD

**Issue by mietek** [agda/agda#4911]
I don’t think I’m the only one who finds `mutual` blocks to be more readable than new-style implicit mutual recursion.  Unfortunately, `mutual` blocks seem to be less expressive.

```agda
module MutualBug where


module Works where
  data Cx : Set

  data Hyp (Γ : Cx) : Set where

  data Cx where
    · : Cx
    _,_ : (Γ : Cx) → Hyp Γ → Cx

  postulate
    wk-hyp : ∀ {Γ A} → Hyp Γ → Hyp (Γ , A)


module DoesNotWork where
  mutual
    data Cx : Set where
      · : Cx
      _,_ : (Γ : Cx) → Hyp Γ → Cx

    data Hyp (Γ : Cx) : Set where

    postulate
      wk-hyp : ∀ {Γ A} → Hyp Γ → Hyp (Γ , A)

{-
/Users/mietek/MutualBug.agda:26,41-42
Not in scope:
  , at /Users/mietek/MutualBug.agda:26,41-42
    (did you mean
       'Works.Cx._,_' or
       'Works._,_'?)
when scope checking ,
-}
```

**Comment by UlfNorell** [comment]
What's your actual proposal here?

The order of checking definitions and type signatures matters. In a new-style mutual block you give them in the order to be checked and in an old-style `mutual` block we check signatures first and then definitions. I don't see how to provide the flexibility of the former while (A) keeping the readability and (B) not breaking all code using old-style mutual.

**Comment by andreasabel** [comment]
The reason why we introduced new-style implicit mutual blocks was exactly to make examples like yours work, @mietek .  While I also like the explicit `mutual` better, it does not cover your example.  

One could think of pragmas that instruct Agda on the order of type-checking the individual declarations and definitions, but this would need a concrete proposal---and if the solution violates esthetics too much, it might not be preferred by users over the implicit mutual blocks.  However, a proposal is welcome!

**Comment by mietek** [comment]
It seems to me that my attempts at using new-style implicit mutual recursion amount to manually maintaining a topological ordering on definitions and type signatures.  Could we have explicit `mutual` blocks automatically maintain this topological ordering?

**Comment by nad** [comment]
Would that change break existing code? Can you give a concrete example of what you have in mind?

**Comment by mietek** [comment]
As I understand, `mutual` blocks can be described as desugaring to new-style implicit mutual recursion in which all the type signatures are given before all the definitions.  I’m wondering if instead `mutual` blocks could desugar to new-style implicit mutual recursion in which the definitions and type signatures are given ordered according to their dependencies.  This change seems conservative in the sense that it should only _unbreak_ existing broken code.

In my example above, the `module Works` contains definitions and type signatures ordered according to their dependencies — that is, `Cx` must be declared before `Hyp`, and `_,_` must be defined before `wk-hyp`.  This is a [topological ordering](https://en.wikipedia.org/wiki/Topological_sorting) that I’ve computed myself.  I’d like to avoid having to compute such orderings myself, especially in more complex scenarios.

**Comment by andreasabel** [comment]
I think this proposal is reasonable, although there is a complication with hidden arguments.  Even if the dependency ordering computed from the visible identifiers is well-founded, the fully reconstructed program might not be well-founded any more.  One cannot necessarily commit on a linearization before the implicit arguments have been computed.  Thus, we have the safe fallback of letting the user declare explicitly the dependency order (and its linearization)---both with explicit and implicit mutual blocks.

A practical obstacle is that mutual blocks are resolved in the nicifier at the moment (`Syntax.Concrete.Definitions`).  A topological order could only be computed by the scope checker that resolves identifiers.  This is not a principle obstacle, though.

A caveat is shadowing.  One should verify that a automatic ordering does not invalidate some programs because identifiers are in scope which would not have been in scope before.

**Comment by andreasabel** [comment]
Here is an example that illustrates ambiguities when a topological order is automatically computed:
```agda
module Issue4911 where

module Explicit where
  mutual
    data C : Set where
      c : D → C

    f : C → Set₁
    f (c d) = Set

    data D : Set where
      d e : D

-- This explicit mutual block has (at least) two possible topological sorts:

-- The first one works, as d on the lhs is treated as a pattern variable.

module Implicit1 where
    data D : Set

    data C : Set where
      c : D → C

    f : C → Set₁
    f (c d) = Set

    data D where
      d e : D

-- The second one fails, as d on the lhs is treated as a constructor.

module Implicit2 where

    data D : Set where
      d e : D

    data C : Set where
      c : D → C

    f : C → Set₁
    f (c d) = Set

-- Incomplete pattern matching for f. Missing cases:
--   f (c e)
-- when checking the definition of f
```
One would have to make a choice:
1. Either the identifiers in patterns are not included when computing the topological order.  This might often lead to undesired (and hard to sell) results.
2. Or the identifiers in patterns are included.  Then the above explicit `mutual` block would be rejected, which would not be backwards compatible.

Initially, when Agda was designed, the question was on the table whether to imitate Haskell's "by default everything is mutual" along the lines you are suggesting (use topological sorting).  However, this was discarded mostly because of hidden arguments.

I am tending to closing this discussion as `wont-fix`.

**Comment by mietek** [comment]
In your example, `Implicit2` seems strongly preferable to `Implicit1`, because reusing the same identifiers for constructors and variables is more likely to happen by mistake than by choice.  In any case, I will accept your decision to close the discussion, should you decide to do so.

**Comment by andreasabel** [comment]
> In your example, `Implicit2` seems strongly preferable to `Implicit1`, because reusing the same identifiers for constructors and variables is more likely to happen by mistake than by choice.

I agree, and it would be good if Agda would warn about pattern variable `d` (as it does in other cases).  However, to warn about `d` would require to maintain a structure for "what the user could think that is in scope" in addition to "what Agda thinks that is in scope" with the former being a superset of the latter.  In a explicit mutual block, it is clear what this over-approximation of the scope should be: all identifiers declared in the mutual block.  In an implicit mutual block, this overapproximation would be the actual scope, I suppose.
